README.md

## XrmToolBoxWeb

XrmToolBoxWeb is a Web application that connects to Microsoft Dataverse.

Idea is to create an "alter ego" of amazing window tool [XrmToolBox](https://www.xrmtoolbox.com)

## Supported Features

Feature | Description | Documentation
 --- | --- | ---
WhoAmI | Retrieves the system user ID for the currently logged on user or the user under whose context the code is running. | [link](https://docs.microsoft.com/en-us/dynamics365/customer-engagement/web-api/whoami?view=dynamics-ce-odata-9)
RetrieveAllEntities | Retrieves metadata information about all the entities. | [link](https://docs.microsoft.com/en-us/dynamics365/customer-engagement/web-api/retrieveallentities?view=dynamics-ce-odata-9) 
RetrieveTotalRecordCount | Returns data on the total number of records for specific entities. | [link](https://docs.microsoft.com/en-us/dynamics365/customer-engagement/web-api/retrievetotalrecordcount?view=dynamics-ce-odata-9) 
GlobalOptionSetDefinitions | fetch Global OptionSets. | [link](https://crmtipoftheday.com/532/retrieving-global-optionsets-using-web-api/) 
System Forms | Gets a collection of SystemForm entity references. |  [link](https://docs.microsoft.com/en-us/dynamics365/customer-engagement/web-api/systemform?view=dynamics-ce-odata-9) 
Calculate Rollup Field | Calculates the value of a rollup attribute. | [link](https://docs.microsoft.com/en-us/previous-versions/dynamicscrm-2016/developers-guide/mt718083(v=crm.8)?redirectedfrom=MSDN)
Commandbar Elements | List(Delete) Commandbar Dependencies | [link](https://taerimhan.com/quick-fix-cant-delete-command-library-component/) 


## Reference

* [Xrm-WebApi-Client project](https://github.com/XRM-OSS/Xrm-WebApi-Client)
* [Microsoft Authentication Library for JavaScript (MSAL.js) 2.0 for Browser-Based Single-Page Applications](https://www.npmjs.com/package/@azure/msal-browser)
* [Acquiring an OAuth Access Token from Dynamics 365 CRM Online](https://giangpham.io/blog/acquiring-an-oauth-access-token-from-dynamics-365-crm-online/)
* [Web API Reference](https://docs.microsoft.com/en-us/dynamics365/customer-engagement/web-api/about?view=dynamics-ce-odata-9)


CHANGELOG.md
## Unreleased

### Fix

- **local-pkg**: add react and msal as local deps to freeze their versions
- fix oauth scopes management

### Feat

- **GlobalOptionSetDefinitions**: add notebook export file
- removed clien-main module and reused the original xrmtoolboxweb module. #4
- integrate jsnotebook
- integrate jsnotebook
- **callaction**: implement call without parameters and return
- **callaction**: module bootstrap

### Refactor

- redirect to client-main page
- upgrade jsnotebook packages and refactor related code
- **callaction**: clean code

## v1.2.4 (2021-12-13)

### Refactor

- **name**: from xrmtoolboxweb-core to @bsorrentino/xrmtoolboxweb-core
- **name**: from xrmtoolboxweb-webapiclient to @bsorrentino/xrmtoolboxweb-webapiclient
- **deps**: Use bootstrap hoist + strict options

## v1.2.0 (2021-11-19)

### Feat

- **service**: add CalculateRollupField

### Fix

- **config**: ts config "target": "ES2015" and "module": "es2020"


src/index.ts


import * as WebApiClient from "xrm-webapi-client";
import * as msal from "@azure/msal-browser";
import { DependentComponentType } from './types'

// if using cdn version, 'Msal' will be available in the global scope

const POWERPLATFORM_ENV_URL = 'https://bsc-labs.crm.dynamics.com'

const msalConfig : msal.Configuration = {
    auth: {
        clientId: 'c43d3729-2e9c-4254-bd68-fbbe4a45301b',
        redirectUri: 'http://localhost:1234',
        //authority: "https://login.microsoftonline.com/common",
        //knownAuthorities: [],
        //navigateToLoginRequestUrl: true
    },
    cache: {
        cacheLocation: 'localStorage'
    }
};

const msalInstance = new msal.PublicClientApplication(msalConfig);


const prepareWebApiRequest = ( auth:msal.AuthenticationResult ) => {
    
    let client = WebApiClient as any
    client.Token = auth.accessToken
    client.ClientUrl = POWERPLATFORM_ENV_URL
}

const whoAmIRequest = () => {
    
    WebApiClient.Execute(WebApiClient.Requests.WhoAmIRequest)
        .then( (response:any ) =>  console.log(response) )
        .catch( (error:any) => console.error(error) )

}


const componentId = [
    { 'ObjectId':'73015cdf-d6a6-4700-9bd2-22128b9933e8','ComponentType':DependentComponentType.Workflow},
    { 'ObjectId':'d06de78f-2062-eb11-a812-00224808901c', 'ComponentType':DependentComponentType.Entity },
]


const retrieveDependenciesForDeleteRequest = () => {

    let request = WebApiClient.Requests.RetrieveDependenciesForDeleteRequest.with( {  urlParams: componentId[1] } )
    WebApiClient.Execute( request )
        .then( (response:any ) =>  console.log(response) )
        .catch( (error:any) => console.error(error.message) )

    
}

function getAccount(): msal.AccountInfo | null {
    // need to call getAccount here?
    const currentAccounts = msalInstance.getAllAccounts();
    if (currentAccounts === null) {
        console.log("No accounts detected");
        return null;
    }

    if (currentAccounts.length > 1) {
        // Add choose account code here
        console.log("Multiple accounts detected, need to add choose account code.");
        return currentAccounts[0];
    } else if (currentAccounts.length === 1) {
        return currentAccounts[0];
    }

    return null;
}


function  aquireToken() {
    const accessTokenRequest:msal.PopupRequest = {
        scopes: [ `${POWERPLATFORM_ENV_URL}/user_impersonation` ] 
    }

    const account = getAccount()
    console.log( 'active account', account )
    if( account ) {
        // this avoid popup 
        msalInstance.setActiveAccount( account )
    }

    return msalInstance.acquireTokenSilent( accessTokenRequest )
        .catch( e => {
            console.error( 'acquireTokenSilent error', e )
            return msalInstance.loginPopup(accessTokenRequest) 
        })
}

aquireToken()
.then( prepareWebApiRequest ) 
.then( whoAmIRequest )
.then( retrieveDependenciesForDeleteRequest )
.catch(error => {
    console.error( error )
    alert( error )
});


src/types.ts
/* 
https://docs.microsoft.com/en-us/dynamics365/customer-engagement/web-api/dependency?view=dynamics-ce-odata-9
*/
export enum DependentComponentType	{
    'Entity' = 1,
    'Attribute' = 2,
    'Relationship' = 3,
    'Attribute Picklist Value' = 4,
    'Attribute Lookup Value' = 5,
    'View Attribute' = 6,
    'Localized Label' = 7,
    'Relationship Extra Condition' = 8,
    'Option Set' = 9,
    'Entity Relationship' = 10,
    'Entity Relationship Role' = 11,
    'Entity Relationship Relationships' = 12,
    'Managed Property' = 13,
    'Entity Key' = 14,
    'Privilege' = 16,
    'PrivilegeObjectTypeCode' = 17,
    'Role' = 20,
    'Role Privilege' = 21,
    'Display String' = 22,
    'Display String Map' = 23,
    'Form' = 24,
    'Organization' = 25,
    'Saved Query' = 26,
    'Workflow' = 29,
    'Report' = 31,
    'Report Entity' = 32,
    'Report Category' = 33,
    'Report Visibility' = 34,
    'Attachment' = 35,
    'Email Template' = 36,
    'Contract Template' = 37,
    'KB Article Template' = 38,
    'Mail Merge Template' = 39,
    'Duplicate Rule' = 44,
    'Duplicate Rule Condition' = 45,
    'Entity Map' = 46,
    'Attribute Map' = 47,
    'Ribbon Command' = 48,
    'Ribbon Context Group' = 49,
    'Ribbon Customization' = 50,
    'Ribbon Rule' = 52,
    'Ribbon Tab To Command Map' = 53,
    'Ribbon Diff' = 55,
    'Saved Query Visualization' = 59,
    'System Form' = 60,
    'Web Resource' = 61,
    'Site Map' = 62,
    'Connection Role' = 63,
    'Complex Control' = 64,
    'Field Security Profile' = 70,
    'Field Permission' = 71,
    'Plugin Type' = 90,
    'Plugin Assembly' = 91,
    'SDK Message Processing Step' = 92,
    'SDK Message Processing Step Image' = 93,
    'Service Endpoint' = 95,
    'Routing Rule' = 150,
    'Routing Rule Item' = 151,
    'SLA' = 152,
    'SLA Item' = 153,
    'Convert Rule' = 154,
    'Convert Rule Item' = 155,
    'Hierarchy Rule' = 65,
    'Mobile Offline Profile' = 161,
    'Mobile Offline Profile Item' = 162,
    'Similarity Rule' = 165,
    'Custom Control' = 66,
    'Custom Control Default Config' = 68,
    'Data Source Mapping' = 166,
    'SDKMessage' = 201,
    'SDKMessageFilter' = 202,
    'SdkMessagePair' = 203,
    'SdkMessageRequest' = 204,
    'SdkMessageRequestField' = 205,
    'SdkMessageResponse' = 206,
    'SdkMessageResponseField' = 207,
    'WebWizard' = 210,
    'Index' = 18,
    'Import Map' = 208,
    'Canvas App' = 300,
    'Connector1' = 371,
    'Connector2' = 372,
    'Environment Variable Definition' = 380,
    'Environment Variable Value' = 381,
    'AI Project Type' = 400,
    'AI Project' = 401,
    'AI Configuration' = 402,
    'Entity Analytics Configuration' = 430,
    'Attribute Image Configuration' = 431,
    'Entity Image Configuration' = 432,
    
    }


export enum EntityFilters {
    Entity = 1, // Use this to retrieve only entity information.
    Attributes = 2, // Use this to retrieve entity information plus attributes for the entity.
    Privileges = 4, // Use this to retrieve entity information plus privileges for the entity.
    Relationships = 8, // Use this to retrieve entity information plus entity relationships for the entity.
    All = 15, // Use this to retrieve all data for an entity.
}

src/webapi.ts
import { AuthenticationResult } from "@azure/msal-browser";
import WebApiClient, {  } from "xrm-webapi-client";
const POWERPLATFORM_ENV_URL = 'https://bsc-labs.crm.dynamics.com'

export const scopes = [ `${POWERPLATFORM_ENV_URL}/user_impersonation` ]

export const prepareWebApiRequest = ( auth:AuthenticationResult ) => {
    
    let client = WebApiClient as any
    client.ApiVersion = '9.1'
    client.Token = auth.accessToken
    client.ClientUrl = POWERPLATFORM_ENV_URL
}

////////////////////////////////////////////////////////////////////////////////////////////////////////////
// ENUMERATIONS
////////////////////////////////////////////////////////////////////////////////////////////////////////////

/* 
https://docs.microsoft.com/en-us/dynamics365/customer-engagement/web-api/dependency?view=dynamics-ce-odata-9
*/
export enum DependentComponentType	{
    'Entity' = 1,
    'Attribute' = 2,
    'Relationship' = 3,
    'Attribute Picklist Value' = 4,
    'Attribute Lookup Value' = 5,
    'View Attribute' = 6,
    'Localized Label' = 7,
    'Relationship Extra Condition' = 8,
    'Option Set' = 9,
    'Entity Relationship' = 10,
    'Entity Relationship Role' = 11,
    'Entity Relationship Relationships' = 12,
    'Managed Property' = 13,
    'Entity Key' = 14,
    'Privilege' = 16,
    'PrivilegeObjectTypeCode' = 17,
    'Role' = 20,
    'Role Privilege' = 21,
    'Display String' = 22,
    'Display String Map' = 23,
    'Form' = 24,
    'Organization' = 25,
    'Saved Query' = 26,
    'Workflow' = 29,
    'Report' = 31,
    'Report Entity' = 32,
    'Report Category' = 33,
    'Report Visibility' = 34,
    'Attachment' = 35,
    'Email Template' = 36,
    'Contract Template' = 37,
    'KB Article Template' = 38,
    'Mail Merge Template' = 39,
    'Duplicate Rule' = 44,
    'Duplicate Rule Condition' = 45,
    'Entity Map' = 46,
    'Attribute Map' = 47,
    'Ribbon Command' = 48,
    'Ribbon Context Group' = 49,
    'Ribbon Customization' = 50,
    'Ribbon Rule' = 52,
    'Ribbon Tab To Command Map' = 53,
    'Ribbon Diff' = 55,
    'Saved Query Visualization' = 59,
    'System Form' = 60,
    'Web Resource' = 61,
    'Site Map' = 62,
    'Connection Role' = 63,
    'Complex Control' = 64,
    'Field Security Profile' = 70,
    'Field Permission' = 71,
    'Plugin Type' = 90,
    'Plugin Assembly' = 91,
    'SDK Message Processing Step' = 92,
    'SDK Message Processing Step Image' = 93,
    'Service Endpoint' = 95,
    'Routing Rule' = 150,
    'Routing Rule Item' = 151,
    'SLA' = 152,
    'SLA Item' = 153,
    'Convert Rule' = 154,
    'Convert Rule Item' = 155,
    'Hierarchy Rule' = 65,
    'Mobile Offline Profile' = 161,
    'Mobile Offline Profile Item' = 162,
    'Similarity Rule' = 165,
    'Custom Control' = 66,
    'Custom Control Default Config' = 68,
    'Data Source Mapping' = 166,
    'SDKMessage' = 201,
    'SDKMessageFilter' = 202,
    'SdkMessagePair' = 203,
    'SdkMessageRequest' = 204,
    'SdkMessageRequestField' = 205,
    'SdkMessageResponse' = 206,
    'SdkMessageResponseField' = 207,
    'WebWizard' = 210,
    'Index' = 18,
    'Import Map' = 208,
    'Canvas App' = 300,
    'Connector1' = 371,
    'Connector2' = 372,
    'Environment Variable Definition' = 380,
    'Environment Variable Value' = 381,
    'AI Project Type' = 400,
    'AI Project' = 401,
    'AI Configuration' = 402,
    'Entity Analytics Configuration' = 430,
    'Attribute Image Configuration' = 431,
    'Entity Image Configuration' = 432,
    
    }

export enum EntityFiltersEnum {
    Entity = 'Entity', // Use this to retrieve only entity information.
    Attributes = 'Attributes', // Use this to retrieve entity information plus attributes for the entity.
    Privileges = 'Privileges', // Use this to retrieve entity information plus privileges for the entity.
    Relationships = 'Relationships', // Use this to retrieve entity information plus entity relationships for the entity.
    All = 'All', // Use this to retrieve all data for an entity.
}

////////////////////////////////////////////////////////////////////////////////////////////////////////////
// REQUEST
////////////////////////////////////////////////////////////////////////////////////////////////////////////

export interface WhoAmIResponse {
    BusinessUnitId: string
    OrganizationId: string
    UserId: string
}
export const WhoAmIRequest = () => 
    WebApiClient.Execute<WhoAmIResponse>(WebApiClient.Requests.WhoAmIRequest) as Promise<WhoAmIResponse>

export interface RetrieveAllEntitiesResponse {
    EntityMetadata: Array<Xrm.Metadata.EntityMetadata>
    Timestamp: String
}
export const RetrieveAllEntities = ( params: { EntityFilters:EntityFiltersEnum, RetrieveAsIfPublished:Boolean}) => { 
    
    const req = WebApiClient.Requests.RetrieveAllEntitiesRequest.with( 
        {  urlParams: { EntityFilters:`Microsoft.Dynamics.CRM.EntityFilters'${params.EntityFilters}'`, RetrieveAsIfPublished:false},
            headers: [
                { key:'Access-Control-Allow-Origin', value:'*.crm.dynamics.com'},
                { key:'Access-Control-Allow-Credentials', value:'true' },
                { key:'Access-Control-Allow-Methods', value:'GET, POST, PUT, DELETE, OPTIONS' },
                { key:'Access-Control-Allow-Headers', value:'Origin, Accept, Content-Type, X-Requested-With, X-CSRF-Token' },
            ]
        })
    
    return WebApiClient.Execute<RetrieveAllEntitiesResponse>(req) as Promise<RetrieveAllEntitiesResponse>
}

const retrieveDependenciesForDeleteRequest = ( params: { ObjectId:string, ComponentType:DependentComponentType} ) => 
WebApiClient.Execute<Xrm.Metadata.EntityMetadata>( WebApiClient.Requests.RetrieveDependenciesForDeleteRequest.with( {  urlParams: params } ) )


////////////////////////////////////////////////////////////////////////////////////////////////////////////
// RetrieveTotalRecordCount
////////////////////////////////////////////////////////////////////////////////////////////////////////////


class RetrieveTotalRecordCountRequest extends WebApiClient.Requests.Request {
    method = 'GET'
    name = 'RetrieveTotalRecordCount'

    constructor( EntityNames:Array<string> ) {
        super()
        this.urlParams = { EntityNames:JSON.stringify(EntityNames) }
    }
}

export interface RetrieveTotalRecordCountResponse {
    "@odata.context":string
    EntityRecordCountCollection: {
        Count: number,
        IsReadOnly: boolean,
        Keys: Array<string>
        Values: Array<number>
    }
}


export const RetrieveTotalRecordCount = ( EntityNames:Array<string> ) => {

    const req = new RetrieveTotalRecordCountRequest( EntityNames )
    return WebApiClient.Execute<RetrieveTotalRecordCountResponse>( req ) as Promise<RetrieveTotalRecordCountResponse>

}

////////////////////////////////////////////////////////////////////////////////////////////////////////////
// GlobalOptionSetDefinitions
////////////////////////////////////////////////////////////////////////////////////////////////////////////

export namespace GlobalOptionSetDefinitions {

    class Request extends WebApiClient.Requests.Request {
        method = 'GET'
        name = 'GlobalOptionSetDefinitions'
    
        constructor() {
            super()
        }

        buildUrl() {
            return super.buildUrl()
            // + '?$select=Name,OptionSetType'
        }
    }

    export type OptionSetType = 'Picklist' | 'Boolean' | string

    export type Metadata = {
        "@odata.type":          '#Microsoft.Dynamics.CRM.OptionSetMetadata',
        ParentOptionSetName:    string|null,
        IsCustomOptionSet:      boolean,
        IsGlobal:               boolean,
        IsManaged:              boolean,
        Name:                   string,
        ExternalTypeName:       string|null,
        OptionSetType:          OptionSetType,
        IntroducedVersion:      string,
        MetadataId:             string,
        HasChanged:             boolean|null,
        Options:                Array<Xrm.Metadata.OptionMetadata>  
    }

    export type Response = { 
        "@odata.context":string 
        value:Array<Metadata>
    } 
        
    export const Invoke = () => {

        const req = new Request()
        return WebApiClient.Execute<Response>( req ) as Promise<Response>
    
    }
}




